Fix empty submobjects distorting width/height of parent Mobject

gvauvert opened the pull request:
## Overview: What does this pull request change?
<!--changelog-start-->
Fixes #4087

When a Mobject contains an empty VGroup, its width or height can be wrong
<!--changelog-end-->

## Motivation and Explanation: Why and how do your changes improve the library?
Bug fix + tests

## Links to added or changed documentation pages

No documentation page, bug fix only.

## Further Information and Comments
* add test
* fix bug
* fix another test which result depends on the bug

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

gvauvert committed:
Issue-4087 Add bug test

gvauvert committed:
Issue-4087 Fix bug

gvauvert committed:
Issue-4087 Fix comment and add type hint

gvauvert committed:
Issue-4087 Fix comment

gvauvert committed:
Issue-4087 Improve previous test instead of adding a new one

gvauvert commented on tests/module/mobject/mobject/test_mobject.py:
```diff
@@ -136,21 +136,19 @@ def test_mobject_dimensions_nested_mobjects():
 
 
 def test_mobject_dimensions_mobjects_with_no_points_are_at_origin():
```
As the previous test was testing an unexpected behavior, it should have been linked to an issue.

chopan050 replied:
I agree.

Now, given that the unexpected behavior is fixed and empty Mobjects are no longer considered as "at the origin", it would be nice to change the name of this test in order to reflect that:
```suggestion
def test_mobject_dimensions_mobjects_with_no_points():
```

gvauvert commented on manim/mobject/mobject.py:
```diff
@@ -2149,10 +2153,11 @@ def get_nadir(self) -> Point3D:
 
     def length_over_dim(self, dim: int) -> float:
```
As this function is meant to be used only internally, it should be private
```suggestion
    def __length_over_dim(self, dim: int) -> float:
```
or at least protected
```suggestion
    def _length_over_dim(self, dim: int) -> float:
```

chopan050 replied:
It makes sense. In that case, `.reduce_across_dimension()` is an even stronger case of a method which should be marked as internal.

There's already a PR which proposes marking some methods as "internal". It's worth checking it:
https://github.com/ManimCommunity/manim/pull/3793

chopan050 commented on manim/mobject/mobject.py:
```diff
@@ -1979,13 +1979,13 @@ def restore(self) -> Self:
         return self
 
-    def reduce_across_dimension(self, reduce_func: Callable, dim: int):
+    def reduce_across_dimension(self, reduce_func: Callable, dim: int) -> float | None:
         """Find the min or max value from a dimension across all points in this and submobjects."""
```
As this method can now also return `None`, it would be nice to document this new behavior in the docstring. Something along the lines of:
```suggestion
    def reduce_across_dimension(
        self, reduce_func: Callable[[Iterable[float]], float], dim: int
    ) -> float | None:
        """Find the min or max value from a dimension across all points in this Mobject and its
        submobjects. This allows for using :meth:`~.length_over_dim` to calculate its length over
        a dimension, i.e. its height, width or depth. If this Mobject is empty, return ``None``,
        since this Mobject should not be taken into account when calculating lengths.
        
        Parameters
        ----------
        reduce_func
            The reducer function to use in order to calculate a value over a dimension.
        dim
            The dimension to use. It should be 0, 1 or 2, representing the X, Y or Z coordinate,
            respectively.
            
        Returns
        -------
        float | None
            The min or max value over the dimension specified by ``dim``, or ``None`` if this
            Mobject is empty.     	
        """
```

chopan050 reviewed (changes requested):
Thanks for this fix! I left some suggestions:

chopan050 committed:
Merge branch 'main' into fix-PR-4087

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

chopan050 committed:
Update tests/module/mobject/mobject/test_mobject.py

chopan050 committed:
Enhance docstring for reduce_across_dimension method

Updated docstring for reduce_across_dimension method to clarify its purpose and parameters.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

chopan050 committed:
Remove float typehints for coords in length_over_dim

chopan050 committed:
Modify test for empty VMobject dimensions

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
